### PR TITLE
KKRT Proper Pseudorandom Encode and Precompute Hash in Stage 1

### PR DESCRIPTION
**Pull request by xcapaldi**
* PseudorandomCode now uses a Highway Hasher to encode the full ID along with its hash index into each AES block.
* Protocol is restructured so the sender can precompute hashes in stage 1 while the receiver is populating its Cuckoo hash table

### CODE REVIEW

File: internal/crypto/cipher.go [hunk 1]
```diff
@@ -24,69 +25,53 @@ const nonceSize = 12 //aesgcm NonceSize
 
 // PseudorandomCode is implemented as follows:
 // C(x) = AES(x[:16]) || AES(x[16:32]) || AES(x[32:48]) || AES(x[48:])
```
Comment (juanli16) [inline internal/crypto/cipher.go:27]: Update this part.
```
C(x) = AES(1||x) || AES(2||x) || AES(3||x) || AES(4||x)
```

File: internal/oprf/oprf.go [hunk 2]
```diff
@@ -37,18 +36,16 @@ func NewOPRF(m, baseOT int) (OPRF, error) {
...
+	s []byte   // secret choice bits
+	q [][]byte // m x k bit matrice
 }
 
 // Encode computes and returns OPRF(k, in)
-func (k Key) Encode(j uint64, in []byte, hIdx uint8) (out []byte, err error) {
-	// compute q_i ^ (C(r) & s)
-	out = crypto.PseudorandomCodeWithHashIndex(k.block, in, hIdx)
+func (k Key) Encode(j uint64, pseudorandomBytes []byte) (out []byte, err error) {
+	out = pseudorandomBytes
```
Comment (juanli16) [inline internal/oprf/oprf.go:45]: you can just change `pseudorandomBytes` without declaring `out`, we won't need that value anymore once it is encoded.

Comment (xcapaldi) [reply internal/oprf/oprf.go]: Okay, done. It speeds up each encode by 2 ns

**Comment by xcapaldi** [comment]
PseudorandomCode now uses TWMB's Murmur3 Hasher to encode the full ID along with its hash index into each AES block.
Protocol is restructured so the sender can precompute hashes in stage 1 while the receiver is populating its Cuckoo hash table.
These combined changes result in performance nearly identical to when no hashing was done in the PseudorandomCode.

File: internal/oprf/improved_kkrt.go [hunk 3]
```diff
@@ -110,26 +98,19 @@ func (ext imprvKKRT) Send(rw io.ReadWriter) (keys Key, err error) {
...
 
 	// store oprf keys
-	aesBlock, err := aes.NewCipher(sk)
-	return Key{block: aesBlock, s: s, q: q}, err
+	return Key{s: s, q: q}, err
 }
 
 // Receive returns the OPRF output on receiver's choice strings using OPRF keys
-func (ext imprvKKRT) Receive(choices *cuckoo.Cuckoo, rw io.ReadWriter) (encodings [cuckoo.Nhash]map[uint64]uint64, err error) {
+func (ext imprvKKRT) Receive(choices *cuckoo.Cuckoo, sk []byte, rw io.ReadWriter) (encodings [cuckoo.Nhash]map[uint64]uint64, err error) {
```
Comment (juanli16) [inline internal/oprf/improved_kkrt.go:108]: no need for `sk` anymore right?

File: pkg/kkrtpsi/sender.go [hunk 4]
```diff
@@ -27,9 +29,9 @@ type Sender struct {
 
 // hashable stores the possible bucket
 // indexes in the receiver cuckoo hash table
-type hashable struct {
-	identifier []byte
-	bucketIdx  [cuckoo.Nhash]uint64
+type pseudorandBytes struct {
```
Comment (juanli16) [inline pkg/kkrtpsi/sender.go:32]: `oprfEncodeInputs`

**Pull request merged**